### ISSUE TITLE
Multi-server setup

### ISSUE THREAD

**Issue by jaredbischof** [jupyterhub/jupyterhub#109]
Am I correct in assuming that there are long-term plans to support multiple jupyterhub servers running as a single instance?  I don't believe that this is possible right now unless I am missing something (please correct me if I'm wrong).  It would be nice to have this for load balancing.  Cheers!

**Comment by yuvipanda** [comment]
I'm going to give it a shot over the next few weeks for jupyter.wmflabs.org, possibly using Docker Swarm. Will keep you guys posted on how it goes.

**Comment by minrk** [comment]
@yuvipanda great, thanks! You might look at compmodels/jupyterhub-deploy, where @jhamrick is using swarm to distribute user containers. That's not what is described here, though, which is multiple Hubs.

**Comment by yuvipanda** [comment]
(Many moons later...)

So I've finally managed to set one up on https://tools.wmflabs.org/paws/hub/oauth_login, with a kubernetes backend. However, the jupyterhub instance itself (+ proxy) are running only once, so it's a SPOF. 

So the two components that exist in the one 'hub' pod now are: The proxy and the jupyterhub itself. Am I right in assuming that if I can somehow synchronize state between all the proxies, and use a mysql/postgres backend for jupyter, I can scale both of these separately however horizontally I want? Is there any useful state in the jupyterhub process itself that isn't stored in the db?

If this is correct, I can probably work on a way to horizontally scale the proxy out, which should work... There are multiple ways to do this, from fanning out to all outputs via a wrapper vs implementing a different proxy with a compatible interface that uses etcd or something to sync data (that people who have more complex setups can use). But if jupyterhub itself stores state, we need to factor that out first...

**Comment by yuvipanda** [comment]
This also helps solve my other problem, which is availability. I like having at least two of everything so I can drain one of traffic and do stuff to it...

So as questions:

1. What state (if any?) is kept in the jupyterhub process itself?
2. If the answer to (1) is 'None', will just providing a scalable proxy be enough?
3. If the answer to (1) is *not* 'None' - what is the state that's kept in there? 

I'm super interested in pushing this forward :)

**Comment by minrk** [comment]
The Hub process can be killed and resumed while leaving all other processes up, so there isn't any long-term state that resides in the Hub. All state is meant to reside in the database. #185 is probably the best illustration of state that resides in the process—mainly transients, such as spawn_pending, etc. I *doubt* it would behave properly if you made two simultaneous spawn requests of the same user on different Hubs using the same database. However, you should be able to do failover - start a second Hub and migrate URL handling before taking down the first Hub.

**Comment by yuvipanda** [comment]
Ok, so I'll try and get the proxy to be distributable by putting some work into it this week and see how it goes!

**Comment by yuvipanda** [comment]
Not quite the same, but somewhat related - I now have a nginx-based Configurable HTTP proxy that jupyterhub can easily use (https://github.com/yuvipanda/jupyterhub-nginx-chp) - it just implements most of the swagger spec and all the jupyterhub functionality I tested works fine. When I hit limits of that, we can probably write another one that scales better across multiple machines.

**Comment by willingc** [comment]
Good information by @yuvipanda, @ssanderson, and others related to this issue. As the issue is more than a year old and I'm not seeing a specific next action, I'm going to close this and mark it as "reference" so it will be discoverable and possibly included in future documentation. Thanks!

**Comment by minrk** [comment]
To run JupyterHub with multiple instances behind a load-balancer, you would have to ensure that the load balancer sends requests for the same user to the same Hub instance every time.

**Comment by jsill14** [comment]
@minrk  Do you need to ensure the load balancer sends the requests for the same user to the same Hub instance to ensure their data is there? If the data was stored or replicated across the Hubs could you spawn the users on any hub?

**Comment by minrk** [comment]
Since a user's server is persistent, you have to make sure that the same Hub gets every request for a given user, at least while that user's server is running in order to route requests properly. This is in-memory state for the Hub, so it isn't shared across instances. The Hub can reconstruct this from information persisted to the database, and does this at startup, but it doesn't reconstruct the state on every request, which would be needed for a single user to be handled correctly across multiple Hubs.

**Comment by jakepaw** [comment]
I am doing something similar. I have 4 servers with jupyterhub running as a service, and are inside of an F5 pool. When i contact jupyterhub.company.com I am routed to one of them to log in. Once I login i am prompted to log in again. and again. and again. Here are my configs:

import sys

c.JupyterHub.admin_access = True
c.JupyterHub.cleanup_proxy = False
c.JupyterHub.cleanup_servers = True
c.Authenticator.admin_users = {'my_user'}
c.Authenticator.enable_auth_state = False
c.Authenticator.post_auth_hook = None

c.SystemdSpawner.default_shell = '/bin/bash'
c.SystemdSpawner.mem_limit = '5G'
c.SystemdSpawner.cpu_limit = 4.0
c.SystemdSpawner.user_workingdir = '/users/{USERNAME}'
c.SystemdSpawner.unit_name_template = 'jupyter-{USERNAME}-singleuser'

c.NotebookApp.terminals_enabledBool = True
c.NotebookApp.terminado_settings= {'shell_command': ['/bin/bash']}
c.NotebookApp.shutdown_no_activity_timeout = 60 * 60

c.JupyterHub.hub_ip = 'my_server_ip'
c.JupyterHub.debug_proxy = True
c.JupyterHub.proxy_api_ip = 'my_server_ip'
c.JupyterHub.spawner_class = 'systemdspawner.SystemdSpawner'
c.JupyterHub.hub_connect_ip = 'my_server_ip'
c.JupyterHub.bind_url = 'https://hostname.company.com:8002'
c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
c.JupyterHub.admin_access = True
c.JupyterHub.ssl_cert = 'cert.pem'
c.JupyterHub.ssl_key = 'key.pem'
c.JupyterHub.services = [
    {
        'name': 'cull-idle',
        'admin': True,
        'command': [sys.executable, 'cull_idle_servers.py', '--timeout=3600'],
    }
]
c.JupyterHub.template_vars = {'announcement': 'This is a temporary Jupyterhub Instance. Send feedback to blank_blank@blank.blank'}


c.LDAPAuthenticator.server_address = 'ldaps://nhsjshshs
c.LDAPAuthenticator.server_port = 636
c.LDAPAuthenticator.bind_dn_template = ['cn={username},ou=it,ou=internal,dc=company,dc=company,dc=com', 'cn={username},ou=health,ou=internal,dc=company,dc=company,dc=com']
c.LDAPAuthenticator.valid_username_regex = '^[anAN][0-9]+$'
c.LDAPAuthenticator.escape_userdn = False
c.LDAPAuthenticator.allowed_groups = None
c.LDAPAuthenticator.attributes = ['cn', 'uidNumber']
c.LDAPAuthenticator.user_attribute = 'sAMAccountName'
c.LDAPAuthenticator.lookup_dn_user_dn_attribute = 'CN'
c.LDAPAuthenticator.allowed_groups = ['cn=hdpeng,ou=it,ou=internal,dc=company,dc=company,dc=com', 'cn=hdpruser,ou=it,ou=internal,dc=company,dc=company,dc=com', 'cn=hdpgpu,ou=it,ou=internal,dc=company,dc=company,dc=com']
